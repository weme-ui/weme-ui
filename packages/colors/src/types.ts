export type ColorAppearance = 'light' | 'dark'

export type ColorType = 'accent' | 'neutral'

export type ColorScalesArray<T = string> = [T, T, T, T, T, T, T, T, T, T, T, T]

export type Colors<T = string> = Record<string, ColorScalesArray<T>>
