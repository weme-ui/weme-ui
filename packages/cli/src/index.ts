#!/usr/bin/env node
import { runMain } from 'citty'
import { cli } from './cli'

runMain(cli)
