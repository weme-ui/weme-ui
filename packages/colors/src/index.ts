export * from './colors'
export * from './transform'
export * from './types'
export * from './utils'
