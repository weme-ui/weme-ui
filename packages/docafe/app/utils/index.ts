export function getUseValue<T = any>(...values: T[]): T {
  let result: T | null = null

  for (const value of values) {
    if (value !== undefined && value !== null) {
      result = value
      break
    }
  }

  return result as T
}
