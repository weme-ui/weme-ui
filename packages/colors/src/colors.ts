import type Color from 'colorjs.io'
import type { ColorAppearance, Colors } from './types'
import { getColorNames, getRadixColor } from './utils'

const lightColors = Object.fromEntries(
  getColorNames().map((name: string) => [
    name,
    getRadixColor('light', name),
  ]),
) as Colors<Color>

const darkColors = Object.fromEntries(
  getColorNames().map((name: string) => [
    name,
    getRadixColor('dark', name),
  ]),
) as Colors<Color>

const lightNeutralColors = Object.fromEntries(
  getColorNames('neutral').map((name: string) => [
    name,
    getRadixColor('light', name),
  ]),
) as Colors<Color>

const darkNeutralColors = Object.fromEntries(
  getColorNames('neutral').map((name: string) => [
    name,
    getRadixColor('dark', name),
  ]),
) as Colors<Color>

export function getColors(appearance: ColorAppearance, isNeutral?: boolean) {
  return isNeutral
    ? appearance === 'light' ? lightNeutralColors : darkNeutralColors
    : appearance === 'light' ? lightColors : darkColors
}
