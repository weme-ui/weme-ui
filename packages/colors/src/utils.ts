import type { ColorAppearance, ColorScalesArray, ColorType } from './types'
import * as RadixColors from '@radix-ui/colors'
import Color from 'colorjs.io'

const SPECIAL_COLOR_NAMES = ['black', 'white'] as const
const NEUTRAL_COLOR_NAMES = ['gray', 'mauve', 'slate', 'sage', 'olive', 'sand'] as const
const ACCENT_COLOR_NAMES = ['tomato', 'red', 'ruby', 'crimson', 'pink', 'plum', 'purple', 'violet', 'iris', 'indigo', 'blue', 'cyan', 'teal', 'jade', 'green', 'grass', 'bronze', 'gold', 'brown', 'orange', 'amber', 'yellow', 'lime', 'mint', 'sky'] as const
const COLOR_NAMES = [...SPECIAL_COLOR_NAMES, ...NEUTRAL_COLOR_NAMES, ...ACCENT_COLOR_NAMES] as const

export type RadixColorNames = keyof typeof RadixColors

export type AllColorNames = typeof NEUTRAL_COLOR_NAMES[number] | typeof ACCENT_COLOR_NAMES[number]
export type AccentColorNames = typeof ACCENT_COLOR_NAMES[number]
export type NeutralColorNames = typeof NEUTRAL_COLOR_NAMES[number]

export function isRadixColorName(color: string) {
  return COLOR_NAMES.includes(color as any)
}

export function getColorNames(type?: ColorType) {
  return !type
    ? COLOR_NAMES
    : type === 'accent'
      ? ACCENT_COLOR_NAMES
      : NEUTRAL_COLOR_NAMES
}

export function getRadixColorName(name: string, appearance: ColorAppearance) {
  return name === 'black'
    ? 'blackP3A'
    : name === 'white'
      ? 'whiteP3A'
      : `${name}${appearance === 'light' ? 'P3' : 'DarkP3'}` as RadixColorNames
}

export function getRadixColor<
  S extends boolean = false,
  T extends S extends true ? string : Color = S extends true ? string : Color,
  RT = ColorScalesArray<T>,
>(
  appearance: ColorAppearance,
  name: string,
  stringify?: boolean,
): RT {
  const colorName = getRadixColorName(name, appearance)

  return Object.values(RadixColors[colorName]).map(
    color => stringify
      ? toColorString(color)
      : new Color(color).to('oklch'),
  ) as RT
}

export function toColorString(colorOrStr: Color | string): string {
  const color = typeof colorOrStr === 'string'
    ? new Color(colorOrStr)
    : colorOrStr

  return color
    .to('oklch')
    .toString({ precision: 4 })
}
