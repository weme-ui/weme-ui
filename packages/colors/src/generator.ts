import type { Colors, ColorScalesArray } from './types'
import BezierEasing from 'bezier-easing'
import Color from 'colorjs.io'
import { getColorNames, toColorString } from './utils'

export function generateColor(
  color: string,
  colors: Colors<Color>,
  background: string,
): ColorScalesArray {
  const sourceColor = new Color(color).to('oklch')
  const backgroundColor = new Color(background).to('oklch')

  const scales = getScaleColors(sourceColor, colors, backgroundColor)

  scales[8] = sourceColor
  scales[9] = getHoverColor(sourceColor, [scales])

  scales[10].coords[1] = Math.min(
    Math.max(scales[8].coords[1], scales[7].coords[1]),
    scales[10].coords[1],
  )

  scales[11].coords[1] = Math.min(
    Math.max(scales[8].coords[1], scales[7].coords[1]),
    scales[11].coords[1],
  )

  return scales.map(toColorString) as ColorScalesArray
}

function getScaleColors(
  source: Color,
  colors: Record<string, ColorScalesArray<Color>>,
  backgroundColor: Color,
): ColorScalesArray<Color> {
  const allColors: { name: string, color: Color, distance: number }[] = []

  Object.entries(colors).forEach(([name, scale]) => {
    for (const color of scale) {
      const distance = source.deltaEOK(color)
      allColors.push({ name, distance, color })
    }
  })

  allColors.sort((a, b) => a.distance - b.distance)

  const closestColors = allColors.filter(
    (color, i, arr) => i === arr.findIndex(value => value.name === color.name),
  )

  const grayColorNames = getColorNames('neutral') as readonly string[]
  const allAreGrays = closestColors.every(color =>
    grayColorNames.includes(color.name),
  )
  if (!allAreGrays && grayColorNames.includes(closestColors[0].name)) {
    while (grayColorNames.includes(closestColors[1].name)) {
      closestColors.splice(1, 1)
    }
  }

  const closestA = closestColors[0]
  const closestB = closestColors[1]

  const a = closestB.distance
  const b = closestA.distance
  const c = closestA.color.deltaEOK(closestB.color)

  const cosA = (b ** 2 + c ** 2 - a ** 2) / (2 * b * c)
  const radA = Math.acos(cosA)
  const sinA = Math.sin(radA)

  const cosB = (a ** 2 + c ** 2 - b ** 2) / (2 * a * c)
  const radB = Math.acos(cosB)
  const sinB = Math.sin(radB)

  const tanC1 = cosA / sinA
  const tanC2 = cosB / sinB
  const ratio = Math.max(0, tanC1 / tanC2) * 0.5

  const colorA = colors[closestA.name]
  const colorB = colors[closestB.name]

  const scale = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11].map(i =>
    new Color(Color.mix(colorA[i], colorB[i], ratio)).to('oklch'),
  ) as ColorScalesArray<Color>

  const baseColor = scale.slice()
    .sort((a, b) => source.deltaEOK(a) - source.deltaEOK(b))[0]

  const ratioC = source.coords[1] / baseColor.coords[1]

  scale.forEach((color) => {
    color.coords[1] = Math.min(
      source.coords[1] * 1.5,
      color.coords[1] * ratioC,
    )
    color.coords[2] = source.coords[2]
  })

  // Light
  if (scale[0].coords[0] > 0.5) {
    const lightnessScale = scale.map(({ coords }) => coords[0])
    const backgroundL = Math.max(0, Math.min(1, backgroundColor.coords[0]))
    const newLightnessScale = transposeProgressionStart(
      backgroundL,
      [1, ...lightnessScale],
      [0, 2, 0, 2] as [number, number, number, number],
    )

    newLightnessScale.shift()

    newLightnessScale.forEach((lightness, i) => {
      scale[i].coords[0] = lightness
    })

    return scale
  }

  // Dark
  const ease: [number, number, number, number] = [1, 0, 1, 0]
  const referenceBackgroundColorL = scale[0].coords[0]
  const backgroundColorL = Math.max(0, Math.min(1, backgroundColor.coords[0]))

  const ratioL = backgroundColorL / referenceBackgroundColorL

  if (ratioL > 1) {
    const maxRatio = 1.5

    for (let i = 0; i < ease.length; i++) {
      const metaRatio = (ratioL - 1) * (maxRatio / (maxRatio - 1))
      ease[i] = ratioL > maxRatio ? 0 : Math.max(0, ease[i] * (1 - metaRatio))
    }
  }

  const lightnessScale = scale.map(({ coords }) => coords[0])
  const newLightnessScale = transposeProgressionStart(
    backgroundColorL,
    lightnessScale,
    ease,
  )

  newLightnessScale.forEach((lightness, i) => {
    scale[i].coords[0] = lightness
  })

  return scale
}

function getHoverColor(source: Color, scales: ColorScalesArray<Color>[]) {
  const [L, C, H] = source.coords
  const newL = L > 0.4 ? L - 0.03 / (L + 0.1) : L + 0.03 / (L + 0.1)
  const newC = L > 0.4 && !Number.isNaN(H) ? C * 0.93 + 0 : C
  const hoverColor = new Color('oklch', [newL, newC, H])

  let closestColor = hoverColor
  let minDistance = Infinity

  scales.forEach((scale) => {
    for (const color of scale) {
      const distance = hoverColor.deltaEOK(color)
      if (distance < minDistance) {
        minDistance = distance
        closestColor = color
      }
    }
  })

  hoverColor.coords[1] = closestColor.coords[1]
  hoverColor.coords[2] = closestColor.coords[2]

  return hoverColor
}

function transposeProgressionStart(
  to: number,
  arr: number[],
  curve: [number, number, number, number],
) {
  return arr.map((n, i, arr) => {
    const lastIndex = arr.length - 1
    const diff = arr[0] - to
    const fn = BezierEasing(...curve)
    return n - diff * fn(1 - i / lastIndex)
  })
}
