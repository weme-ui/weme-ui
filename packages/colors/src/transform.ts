import type { ColorAppearance, ColorScalesArray } from './types'
import { getColors } from './colors'
import { generateColor } from './generator'
import { getRadixColor, isRadixColorName } from './utils'

interface TransformColorOptions {
  color: string
  appearance?: ColorAppearance
  isNeutral?: boolean
  background?: string
}

export function transformColor({
  color,
  appearance = 'light',
  isNeutral = false,
  background = '#fff',
}: TransformColorOptions): ColorScalesArray {
  if (isRadixColorName(color))
    return getRadixColor(appearance, color, true)

  const colors = getColors(appearance, isNeutral)

  return generateColor(color, colors, background)
}
