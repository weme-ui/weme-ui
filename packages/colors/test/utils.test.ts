import { describe, expect, it } from 'vitest'
import { getColorNames, getRadixColor, getRadixColorName, isRadixColorName, toColorString } from '../src/utils'

describe('utils', () => {
  it('should be radix color name', () => {
    [
      ['black', true],
      ['crimson', true],
      ['olive', true],
      ['tomato', true],
      ['bad', false],
    ].forEach(
      ([color, result]) => {
        expect(isRadixColorName(color as string)).toEqual(result)
      },
    )
  })

  it('should get color names', () => {
    expect(getColorNames()).toMatchSnapshot()
    expect(getColorNames('accent')).toMatchSnapshot()
    expect(getColorNames('neutral')).toMatchSnapshot()
  })

  it('should get radix color name', () => {
    [
      ['black', 'light', 'blackP3A'],
      ['white', 'light', 'whiteP3A'],
      ['gray', 'light', 'grayP3'],
      ['tomato', 'light', 'tomatoP3'],
      ['gray', 'dark', 'grayDarkP3'],
      ['tomato', 'dark', 'tomatoDarkP3'],
    ].forEach(
      ([name, appearance, result]) => {
        expect(getRadixColorName(name, appearance as any)).toEqual(result)
      },
    )
  })

  it('should get radix color', () => {
    expect(getRadixColor('light', 'black', true)).toMatchSnapshot()
    expect(getRadixColor('dark', 'olive')).toMatchSnapshot()
  })

  it('should transform color to string', () => {
    expect(toColorString('#06f')).toMatchSnapshot()
    expect(toColorString('color(display-p3 0.733 0.864 0.724)')).toMatchSnapshot()
    expect(toColorString('oklch(0% 0 0)')).toMatchSnapshot()
    expect(toColorString('white')).toMatchSnapshot()
  })
})
