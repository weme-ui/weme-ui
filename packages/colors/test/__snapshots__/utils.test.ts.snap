// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`utils > should get color names 1`] = `
[
  "black",
  "white",
  "gray",
  "mauve",
  "slate",
  "sage",
  "olive",
  "sand",
  "tomato",
  "red",
  "ruby",
  "crimson",
  "pink",
  "plum",
  "purple",
  "violet",
  "iris",
  "indigo",
  "blue",
  "cyan",
  "teal",
  "jade",
  "green",
  "grass",
  "bronze",
  "gold",
  "brown",
  "orange",
  "amber",
  "yellow",
  "lime",
  "mint",
  "sky",
]
`;

exports[`utils > should get color names 2`] = `
[
  "tomato",
  "red",
  "ruby",
  "crimson",
  "pink",
  "plum",
  "purple",
  "violet",
  "iris",
  "indigo",
  "blue",
  "cyan",
  "teal",
  "jade",
  "green",
  "grass",
  "bronze",
  "gold",
  "brown",
  "orange",
  "amber",
  "yellow",
  "lime",
  "mint",
  "sky",
]
`;

exports[`utils > should get color names 3`] = `
[
  "gray",
  "mauve",
  "slate",
  "sage",
  "olive",
  "sand",
]
`;

exports[`utils > should get radix color 1`] = `
[
  "oklch(0% 0 none / 0.05)",
  "oklch(0% 0 none / 0.1)",
  "oklch(0% 0 none / 0.15)",
  "oklch(0% 0 none / 0.2)",
  "oklch(0% 0 none / 0.3)",
  "oklch(0% 0 none / 0.4)",
  "oklch(0% 0 none / 0.5)",
  "oklch(0% 0 none / 0.6)",
  "oklch(0% 0 none / 0.7)",
  "oklch(0% 0 none / 0.8)",
  "oklch(0% 0 none / 0.9)",
  "oklch(0% 0 none / 0.95)",
]
`;

exports[`utils > should get radix color 2`] = `
[
  {
    "alpha": 1,
    "coords": [
      0.17973246055483877,
      0.004277634496519416,
      128.338255479547,
    ],
    "spaceId": "oklch",
  },
  {
    "alpha": 1,
    "coords": [
      0.21165579405893234,
      0.004105841933120096,
      128.28736060851094,
    ],
    "spaceId": "oklch",
  },
  {
    "alpha": 1,
    "coords": [
      0.2515841298813772,
      0.005082727095308246,
      129.00851914367138,
    ],
    "spaceId": "oklch",
  },
  {
    "alpha": 1,
    "coords": [
      0.2807454069652463,
      0.0056028888898918515,
      131.4360498780601,
    ],
    "spaceId": "oklch",
  },
  {
    "alpha": 1,
    "coords": [
      0.3100843491295378,
      0.006558806521863961,
      131.44715749995157,
    ],
    "spaceId": "oklch",
  },
  {
    "alpha": 1,
    "coords": [
      0.3463972476514622,
      0.0076470079390652925,
      134.43131593036014,
    ],
    "spaceId": "oklch",
  },
  {
    "alpha": 1,
    "coords": [
      0.3985962298858755,
      0.009645027679510223,
      136.02536798373785,
    ],
    "spaceId": "oklch",
  },
  {
    "alpha": 1,
    "coords": [
      0.4876956222394584,
      0.012367493765092231,
      140.77952511590513,
    ],
    "spaceId": "oklch",
  },
  {
    "alpha": 1,
    "coords": [
      0.5352299554471645,
      0.017614383640660522,
      139.61437772386046,
    ],
    "spaceId": "oklch",
  },
  {
    "alpha": 1,
    "coords": [
      0.5812375261552609,
      0.01632615521390139,
      140.06362826061832,
    ],
    "spaceId": "oklch",
  },
  {
    "alpha": 1,
    "coords": [
      0.7659440335244477,
      0.012765356783994604,
      139.49583931218433,
    ],
    "spaceId": "oklch",
  },
  {
    "alpha": 1,
    "coords": [
      0.9470128722237816,
      0.003355782945848951,
      144.7599607460728,
    ],
    "spaceId": "oklch",
  },
]
`;

exports[`utils > should transform color to string 1`] = `"oklch(56.35% 0.2408 260.8)"`;

exports[`utils > should transform color to string 2`] = `"oklch(85.78% 0.0714 146.8)"`;

exports[`utils > should transform color to string 3`] = `"oklch(0% 0 0)"`;

exports[`utils > should transform color to string 4`] = `"oklch(100% 0 none)"`;
